### PR TITLE
OBPIH-6520, OBPIH-6412 Dates localization

### PR DESCRIPTION
**Pull request by alannadolny**
This ticket contains code for:
- OBPIH-6520 Set up for date localization in DD/Month/YYYY format for Es and Mx locales 
(Previously was in: https://github.com/openboxes/openboxes/pull/4682)
- OBPIH-6412 Localize date on inbound workflow
(Previously was in: https://github.com/openboxes/openboxes/pull/4689)

### CODE REVIEW

File: grails-app/views/stockMovement/_summary.gsp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+<%@ page import="org.pih.warehouse.DateFormatName" %>
```
Comment (jmiranda) [inline grails-app/views/stockMovement/_summary.gsp:1]: This should be in a package like "core". Unless there's a better place for it.

Comment (jmiranda) [reply grails-app/views/stockMovement/_summary.gsp:1]: Or "util" if people feel more strongly about that.

File: grails-app/views/stockMovement/_summary.gsp [hunk 2]
```diff
@@ -91,37 +92,60 @@
                         <g:if test="${!stockMovement?.dateRequested }">
                             <span class="dateRequested">
                                 <warehouse:message code="stockMovement.dateRequested.label"/>:
-                                <label><format:date obj="${stockMovement?.dateRequested}"/></label>
+                                <label>
+                                    <g:formatDate
+                                            formatName="${DateFormatName.ABBREVIATED_MONTH.property}"
+                                            date="${stockMovement?.dateRequested}"
```
Comment (jmiranda) [inline grails-app/views/stockMovement/_summary.gsp:98]: I need more time to think about this and review what's currently happening, but I'm going to post the rest of my review as-is.

File: grails-app/i18n/messages.properties [hunk 3]
```diff
@@ -655,6 +654,13 @@ documents.label=Documents
 documentType.label=Document Type
 documentTypes.label=Document Types
 
+# Date formats
+datetime.format=dd/MMM/yyyy hh:mm:ss a z
+date.format=MM/dd/yyyy
+monthyear.format=MM/yyyy
+fullmonth.date.format=MMMM dd, yyyy
+abbreviatedmonth.date.format=dd/MMM/yyyy
```
[the remark this answers is not in the document]
Comment (alannadolny) [reply grails-app/i18n/messages.properties:662]: Basically, I just changed the labels because of the request in this PR: https://github.com/openboxes/openboxes/pull/4682

File: grails-app/i18n/messages.properties [hunk 4]
```diff
@@ -3232,6 +3238,12 @@ react.default.select.typeToSearch.label=Type to search
 react.default.noOptions.label=No options
 react.default.dateInput.placeholder.label=Select a date
 react.default.form.field.required.label=${0} is required
+
+# Date formats
+react.default.datetime.format=MM/DD/YYYY HH:mm:ss Z
+react.default.date.format=MM/DD/YYYY
+react.default.monthyear.format=MM/YYYY
```
[the remark this answers is not in the document]
Comment (alannadolny) [reply grails-app/i18n/messages.properties:3245]: It's necessary because we are pulling only the labels starting with `react`... And the `react.default` are the labels we always pull.
And yes, there is a small change between the formats.

File: grails-app/views/stockMovement/_summary.gsp (hunk 1 above)
Comment (alannadolny) [reply grails-app/views/stockMovement/_summary.gsp:1]: I didn't know where it should be placed, so thanks for this comment.

Comment (jmiranda) [reply grails-app/views/stockMovement/_summary.gsp:1]: Yeah let's move it to `src/main/groovy/org/pih/warehouse/core` since that's where most of the enums are located. And let's change the name of the enum to just DateFormat. Or if that's problematic because of java.text.DateFormat, then maybe DateFormatCode. Something is bothering me about DateFormatName. If anyone has any other suggestions please add it here.

File: grails-app/i18n/messages.properties [hunk 5]
```diff
@@ -655,6 +654,13 @@ documents.label=Documents
 documentType.label=Document Type
 documentTypes.label=Document Types
 
+# Date formats
+datetime.format=dd/MMM/yyyy hh:mm:ss a z
+date.format=MM/dd/yyyy
```
[the remark this answers is not in the document]
Comment (jmiranda) [reply grails-app/i18n/messages.properties:659]: I have confirmed that the order of precedence is: `date.format` > `default.date.format` > 'yyyy-MM-dd HH:mm:ss z' so date.format **would** override default.date.format.

https://github.com/grails/grails-gsp/blob/cae0d8182705bf1481ae53c5e2d81511356e4640/grails-plugin-gsp/src/main/groovy/org/grails/plugins/web/taglib/FormatTagLib.groovy#L181

File: grails-app/views/stockMovement/_summary.gsp (hunk 1 above)
Comment (jmiranda) [reply grails-app/views/stockMovement/_summary.gsp:1]: I'm trying something new with respect to making decisions as a team, so please take this poll when you have a minute.

https://github.com/openboxes/openboxes/discussions/4693

@alannadolny @awalkowiak @EWaterman @drodzewicz @kchelstowski

File: grails-app/i18n/messages.properties [hunk 6]
```diff
@@ -357,7 +357,9 @@ default.credits.label=Credits
 default.balance.label=Balance
 default.dashboard.label=Dashboard
 default.data.label=Data
-default.date.format=dd/MMM/yyyy hh:mm:ss a z
+datetime.format=dd/MMM/yyyy hh:mm:ss a z
+date.format=MM/dd/yyyy
```
Comment (awalkowiak) [inline grails-app/i18n/messages.properties]: This should stay as a `default.date.format` due to order of precedence resolving date formats

**Review by awalkowiak — COMMENTED** [review]
Test review of specific commit

File: grails-app/i18n/messages.properties (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmiranda) [reply grails-app/i18n/messages.properties:659]: Ok, let's do that.

Add this back 
```
default.date.format=dd/MMM/yyyy hh:mm:ss a z
```

And remove datetime.format and date.format.

File: grails-app/i18n/messages.properties (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmiranda) [reply grails-app/i18n/messages.properties:662]: @EWaterman I see your point, but I think we're going to have a problem either way. The format/type also gets unruly when you look at all of the different date formats we're currently using.

What's missing from this discussion is a discovery of the actual date format patterns used and whether we need them all. And if not, how to standardize and consolidate. That is a rather large refactoring that I don't think we have time or energy to undertake that right now. 

> [!IMPORTANT]
> TL;DR And the fact that the initial discovery did not include a discussion on what's currently used and whether/how we should consolidate makes me realize that we shouldn't be going down the path we're heading now. We should take a step back, change all of our hard-coded formats that use MM to use MMM instead and call it a day. I'm asking Kelsey if this is ok for now.

[the remark this answers is not in the document]
Comment (jmiranda) [reply grails-app/i18n/messages.properties:662]: > the fact that the initial discovery did not include a discussion on what's currently used and whether/how we should consolidate

To be clear, this is not necessarily a problem of the discovery, but a combination of the aim of the discovery i.e. the requirements, objectives, open questions). And to a large extent that falls on us (@jmiranda @EWaterman @awalkowiak). We definitely need to be more thorough with expected outcomes and what questions should be explored. And I'm hoping that will be a collaborative and probably iterative process i.e. as discovery is being done, more questions and avenues should arise. We just need to set the discovery on the right path.
